google/cloud/bigquery/table.py

```diff
@@ -1916,36 +1911,47 @@ def to_dataframe(
             self.schema
         )
 
+        # When converting date or timestamp values to nanosecond precision, the result
+        # can be out of pyarrow bounds. To avoid the error when converting to
+        # Pandas, we set the date_as_object or timestamp_as_object parameter to True,
+        # if necessary.
+        date_as_object = not all(
+            self.__can_cast_timestamp_ns(col)
+            for col in record_batch
+            # Type can be date32 or date64 (plus units).
+            # See: https://arrow.apache.org/docs/python/api/datatypes.html
+            if str(col.type).startswith("date")
+        )
+        if date_as_object:
+            default_dtypes = {
+                name: type_
+                for name, type_ in default_dtypes.items()
+                if type_ != _pandas_helpers.date_dtype_name
+            }
+
         # Let the user-defined dtypes override the default ones.
         # https://stackoverflow.com/a/26853961/101923
         dtypes = {**default_dtypes, **dtypes}
 
-        # When converting timestamp values to nanosecond precision, the result
-        # can be out of pyarrow bounds. To avoid the error when converting to
-        # Pandas, we set the timestamp_as_object parameter to True, if necessary.
-        types_to_check = {
-            pyarrow.timestamp("us"),
-            pyarrow.timestamp("us", tz=datetime.timezone.utc),
-        }
-
-        for column in record_batch:
-            if column.type in types_to_check:
-                try:
-                    column.cast("timestamp[ns]")
-                except pyarrow.lib.ArrowInvalid:
-                    timestamp_as_object = True
-                    break
-        else:
-            timestamp_as_object = False
-
-        extra_kwargs = {"timestamp_as_object": timestamp_as_object}
+        timestamp_as_object = not all(
+            self.__can_cast_timestamp_ns(col)
+            for col in record_batch
+            # Type can be timestamp (plus units and time zone).
+            # See: https://arrow.apache.org/docs/python/api/datatypes.html
+            if str(col.type).startswith("timestamp")
+        )
 
         df = record_batch.to_pandas(
-            date_as_object=date_as_object, integer_object_nulls=True, **extra_kwargs
+            date_as_object=date_as_object,
+            timestamp_as_object=timestamp_as_object,
+            integer_object_nulls=True,
         )
 
         for column in dtypes:
-            df[column] = pandas.Series(df[column], dtype=dtypes[column])
+            data = df[column]
+            if dtypes[column] == date_dtype_name:
+                data = DateArray(data.to_numpy(copy=False), copy=False)
```

Comment:
Rather than a special check for the date dtype, I wonder if we can use the `types_mapper` parameter on `to_pandas`?

https://arrow.apache.org/docs/python/generated/pyarrow.RecordBatch.html#pyarrow.RecordBatch.to_pandas

Reply:
Yes, it should be possible, but we'll need some additional work on the db-dtypes package. https://github.com/googleapis/python-db-dtypes-pandas/issues/38